### PR TITLE
Retry transaction failures.

### PR DESCRIPTION
**Pull request by psudheer21**
Currently the failed transactions are not retried and are being
recognized as successes. Make sure that the transaction failures are
retried properly. Also make sure that the failures are not added to the
results.

Reviewers:
Mihnea, Rob

### CODE REVIEW

File: src/com/oltpbenchmark/api/Worker.java [hunk 1]
```diff
@@ -402,11 +402,8 @@ public final void run() {
                 }
             }
             // In case of transaction failures, the end times will not be populated.
-            if (executionState.getEndOperation() == 0) {
-              executionState.setEndOperation(System.nanoTime());
-            }
-            if (executionState.getEndConnection() == 0) {
-              executionState.setEndConnection(System.nanoTime());
+            if (executionState.getEndOperation() == 0 || executionState.getEndConnection() == 0) {
+              continue work;
```
Comment (robertsami) [inline src/com/oltpbenchmark/api/Worker.java:414]: it seems like the behavior now will be that we will retry indefinitely. should we retry N times before calling this a failure?

File: src/com/oltpbenchmark/api/Worker.java [hunk 2]
```diff
@@ -402,11 +402,8 @@ public final void run() {
                 }
             }
             // In case of transaction failures, the end times will not be populated.
-            if (executionState.getEndOperation() == 0) {
-              executionState.setEndOperation(System.nanoTime());
-            }
-            if (executionState.getEndConnection() == 0) {
-              executionState.setEndConnection(System.nanoTime());
+            if (executionState.getEndOperation() == 0 || executionState.getEndConnection() == 0) {
```
Comment (robertsami) [inline src/com/oltpbenchmark/api/Worker.java:413]: maybe we should also add some `executionState.incrementRetryCount()`

Comment (psudheer21) [reply src/com/oltpbenchmark/api/Worker.java:413]: Done.

File: src/com/oltpbenchmark/api/Worker.java (hunk 1 above)
Comment (psudheer21) [reply src/com/oltpbenchmark/api/Worker.java:414]: Done.

File: src/com/oltpbenchmark/api/Worker.java [hunk 3]
```diff
@@ -562,6 +569,8 @@ protected final TransactionExecutionState doWork(boolean measure, SubmittedProce
                     // ------------------
                     } else if (ex.getErrorCode() == 0 && ex.getSQLState() != null && ex.getSQLState().equals("40001")) {
                         // Postgres serialization
+                        totalRetries++;
+                        status = TransactionStatus.RETRY;
```
Comment (robertsami) [inline src/com/oltpbenchmark/api/Worker.java:573]: looks like there are other situations where we maybe want to retry, e.g. L587, based on the comment

Comment (psudheer21) [reply src/com/oltpbenchmark/api/Worker.java:573]: I didn't want to retry unless we have transaction conflicts cuz they should be the most common failures that we should retry. Other failures can be ignored now that we don't count failures now as successes.

File: src/com/oltpbenchmark/api/Worker.java [hunk 4]
```diff
@@ -402,11 +410,8 @@ public final void run() {
                 }
             }
             // In case of transaction failures, the end times will not be populated.
-            if (executionState.getEndOperation() == 0) {
-              executionState.setEndOperation(System.nanoTime());
-            }
-            if (executionState.getEndConnection() == 0) {
-              executionState.setEndConnection(System.nanoTime());
+            if (executionState.getEndOperation() == 0 || executionState.getEndConnection() == 0) {
```
Comment (robertsami) [inline src/com/oltpbenchmark/api/Worker.java:413]: do we ever actually get here? and if so, should we be throwing an exception instead of continuing?

**Review by robertsami — APPROVED** [review]
accepting to unblock, but please look at the two comments/questions

File: src/com/oltpbenchmark/api/Worker.java (hunk 4 above)
Comment (psudheer21) [reply src/com/oltpbenchmark/api/Worker.java:413]: yup.
2 cases,
1.  when we don't get a connection from the Hikari connection pool.
2.  When the transaction fails for any reason.

In those cases, we will come here. 
Now we can continue to execute or fail.. I chose to continue. We can fail here as well.

**Pull request merged**